core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/AdaptiveBroadcastToPartitionedJoin.java

```diff
@@ -0,0 +1,460 @@
+import static io.trino.sql.planner.plan.ExchangeNode.partitionedExchange;
+import static io.trino.sql.planner.plan.JoinNode.DistributionType.PARTITIONED;
+import static io.trino.sql.planner.plan.JoinNode.DistributionType.REPLICATED;
+import static io.trino.sql.planner.plan.JoinType.INNER;
+import static io.trino.sql.planner.plan.JoinType.LEFT;
+import static io.trino.sql.planner.plan.Patterns.Join.left;
+import static io.trino.sql.planner.plan.Patterns.Join.right;
+import static io.trino.sql.planner.plan.Patterns.aggregation;
+import static io.trino.sql.planner.plan.Patterns.exchange;
+import static io.trino.sql.planner.plan.Patterns.join;
+import static io.trino.sql.planner.plan.Patterns.source;
+import static io.trino.sql.planner.plan.SimplePlanRewriter.rewriteWith;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Change the join type from broadcast to partitioned join based on runtime data size stats. This rule is only used
+ * during Adaptive Planning in FTE.
+ * From:
+ *  - Join (distribution: REPLICATED)
+ *      - Partial Aggregation (optional)
+ *          - Remote Exchange (distribution: RoundRobin, absent if source is RemoteSource)
+ *              - left side or RemoteSource (distribution: REPARTITIONED)
+ *      - Partial Aggregation (optional)
+ *          - Local Exchange (optional)
+ *              - Remote Exchange (distribution: REPLICATED, absent if source is RemoteSource)
+ *                  - right side or RemoteSource (distribution: REPLICATED)
+ * To:
+ * - Join (distribution: PARTITIONED)
+ *      - Partial Aggregation (optional)
+ *          - Exchange (distribution: PARTITIONED, either change the existing one or add new if needed)
+ *              - left side or RemoteSource (distribution: REPARTITIONED)
+ *      - Partial Aggregation (optional)
+ *          - Local Exchange (optional)
+ *              - Remote Exchange (distribution: PARTITIONED, either change the existing one or add new if needed)
+ *                  - right side or RemoteSource (distribution: REPARTITION)
+ *
+ * Note:
+ * We will try to not add extra Exchange nodes when possible. For example, if the left side is not
+ * considered completed, and still has Exchange node with RoundRobin distribution, we will simply change the distribution
+ * type of that exchange node to PARTITIONED instead of adding a new Exchange node. Same applies to the right side.
+ *
+ * However, if the node is considered completed (i.e. RemoteSourceNode), we will need to add a new Exchange node
+ * with partitioned distribution. We will try to consider the cost of adding an extra Exchange node while
+ * making this decision.
+ *
+ * TODO: We can make this rule more efficient by using the StreamPropertyDerivation mechanism to determine the
+ *  RemoteExchangeNodes that can be reused instead of adding a new one. For instance, this will be helpful in
+ *  cases where either side of the join has union nodes.
+ */
```

Comment:
```suggestion
/**
 * Change the join type from broadcast to partitioned join based on runtime data size stats. This rule is only used
 * during Adaptive Planning in FTE.
 * <p>
 * <pre>
 * From:
 *  - Join (distribution: REPLICATED)
 *      - Partial Aggregation (optional)
 *          - Remote Exchange (distribution: RoundRobin, absent if source is RemoteSource)
 *              - left side or RemoteSource (distribution: REPARTITIONED)
 *      - Partial Aggregation (optional)
 *          - Local Exchange (optional)
 *              - Remote Exchange (distribution: REPLICATED, absent if source is RemoteSource)
 *                  - right side or RemoteSource (distribution: REPLICATED)
 * To:
 * - Join (distribution: PARTITIONED)
 *      - Partial Aggregation (optional)
 *          - Exchange (distribution: PARTITIONED, either change the existing one or add new if needed)
 *              - left side or RemoteSource (distribution: REPARTITIONED)
 *      - Partial Aggregation (optional)
 *          - Local Exchange (optional)
 *              - Remote Exchange (distribution: PARTITIONED, either change the existing one or add new if needed)
 *                  - right side or RemoteSource (distribution: REPARTITION)
 *</pre>
 *
 * <b>Note:</b>
 * We will try to not add extra Exchange nodes when possible. For example, if the left side is not
 * considered completed, and still has Exchange node with RoundRobin distribution, we will simply change the distribution
 * type of that exchange node to PARTITIONED instead of adding a new Exchange node. Same applies to the right side.
 * <p>
 * However, if the node is considered completed (i.e. RemoteSourceNode), we will need to add a new Exchange node
 * with partitioned distribution. We will try to consider the cost of adding an extra Exchange node while
 * making this decision.
 * <p>
 * TODO: We can make this rule more efficient by using the StreamPropertyDerivation mechanism to determine the
 *       RemoteExchangeNodes that can be reused instead of adding a new one. For instance, this will be helpful in
 *       cases where either side of the join has union nodes.
 */
```